File: app/src/main/java/com/lis/player_java/ui/PlayerFragment.java
package com.lis.player_java.ui;

import android.content.Context;
import android.content.SharedPreferences;
import android.icu.text.SimpleDateFormat;
import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.SeekBar;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;

import com.google.android.exoplayer2.Player;
import com.lis.player_java.R;
import com.lis.player_java.data.model.Thumb;
import com.lis.player_java.databinding.FragmentPlayerBinding;
import com.lis.player_java.di.Injection;
import com.lis.player_java.tool.ImageFun;
import com.lis.player_java.viewModel.PlaybackViewModel;

import java.text.MessageFormat;
import java.text.ParseException;
import java.util.Locale;
import java.util.Objects;
import java.util.concurrent.TimeUnit;

public class PlayerFragment extends Fragment {
    private FragmentPlayerBinding binding;
    private PlaybackViewModel viewModel;


    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        binding = FragmentPlayerBinding.inflate(inflater, container, false);
        viewModel = new ViewModelProvider(
                this, getFactory()
        ).get(PlaybackViewModel.class);

        bindElement();
        return binding.getRoot();
    }


    private void bindElement() {
        viewModel.getDuration().observe(getViewLifecycleOwner(), songDuration -> {
            Toast.makeText(requireContext(), ""+songDuration, Toast.LENGTH_SHORT).show();
            binding.songProgress.setMax(songDuration.intValue());
            binding.songDuration.setText(getStringSongDuration(songDuration));
        });

        ImageFun imageFun = new ImageFun();

        viewModel.getMusicInfo().observe(getViewLifecycleOwner(), musicInfo -> {
            if(musicInfo != null){
                Thumb thumb = null;
                try {
                    thumb = musicInfo.getAlbum().getThumb();

                } catch (NullPointerException e){
                    Log.e("info", Objects.toString(musicInfo));
                }

                if(thumb != null){
                    String image = thumb.getPhoto600();
                    imageFun.setImageToBackground(binding.backgroundImage, image);
                    imageFun.setImage(binding.songImage, image);
                } else {
                    //для установки дефолтной картинки
                    imageFun.setImageToBackground(binding.backgroundImage, R.drawable.song_image);
                    imageFun.setImage(binding.songImage, R.drawable.song_image);
                }

                binding.songName.setText(musicInfo.getTitle());
                binding.songAuthor.setText(musicInfo.getArtist());
            }
        });//заполнение сведений о песне

        viewModel.getPosition().observe(getViewLifecycleOwner(), position -> {
            binding.songPosition.setText(getStringSongDuration(position.longValue()));
            binding.songProgress.setProgress(position.intValue());
        });//установка текущей позиции

        viewModel.getDownloadPosition().observe(getViewLifecycleOwner(), downloadPosition -> {
            int progress;
            if(downloadPosition != 0){
                progress = (int) (binding.songProgress.getMax() * downloadPosition / 100);
            } else {
                progress = 0;
            }
            if(binding.songProgress.getSecondaryProgress() != progress){
                binding.songProgress.setSecondaryProgress(progress);
                //Log.e("download", ""+progress+"|"+downloadPosition);
            }

        });

        viewModel.isPlaying().observe(getViewLifecycleOwner(), isPlaying -> {
            imageFun.setImage(binding.buttonPlayPause,
                    isPlaying ? R.drawable.ic_baseline_pause_24
                            : R.drawable.ic_baseline_play_arrow_24);

        });//установка иконки play\pause кнопки

        viewModel.getRepeatMode().observe(getViewLifecycleOwner(), repeatMode -> {
            switch (repeatMode) {
                case Player.REPEAT_MODE_OFF:
                    imageFun.setImage(binding.buttonLoop, R.drawable.ic_baseline_plus_one_24);
                    break;
                case Player.REPEAT_MODE_ONE:
                    imageFun.setImage(binding.buttonLoop, R.drawable.ic_baseline_loop_24);
                    break;
                case Player.REPEAT_MODE_ALL:
                    imageFun.setImage(binding.buttonLoop, R.drawable.ic_baseline_low_priority_24);
                    break;
            }
        });// установка иконки зацикливания

        viewModel.getCount().observe(getViewLifecycleOwner(), count -> {
            Toast.makeText(requireContext(), ""+ count, Toast.LENGTH_SHORT).show();
        });

        binding.songProgress.setOnSeekBarChangeListener(seekBarSelectProgressListener());

        binding.buttonPlayPause.setOnClickListener(v -> startClickListener());

        binding.buttonNext.setOnClickListener(v -> nextClickListener());
        binding.buttonPrevious.setOnClickListener(v -> prevClickListener());

        binding.buttonLoop.setOnClickListener(this::loopClickListener);
    }
    private int repeatMode = Player.REPEAT_MODE_OFF;

    private void loopClickListener(View v) {
        switch (repeatMode) {
            case Player.REPEAT_MODE_OFF:
                repeatMode = Player.REPEAT_MODE_ONE;
                break;
            case Player.REPEAT_MODE_ONE:
                repeatMode = Player.REPEAT_MODE_ALL;
                break;
            case Player.REPEAT_MODE_ALL:
                repeatMode = Player.REPEAT_MODE_OFF;
                break;
        }
        viewModel.setLoopingState(repeatMode);
    }


    private void startClickListener() {
        boolean isPlaying = Boolean.TRUE.equals(viewModel.isPlaying().getValue());
        Log.e("isPlaying", ""+isPlaying);
        if (isPlaying) {
            viewModel.pause();
        } else {
            viewModel.play();
        }
        isPlaying = Boolean.TRUE.equals(viewModel.isPlaying().getValue());
        Log.e("isPlaying1", ""+isPlaying);
    }

    private void nextClickListener() {
        viewModel.nextSong();
    }

    private void prevClickListener() {
        viewModel.prevSong();
    }

    private SeekBar.OnSeekBarChangeListener seekBarSelectProgressListener() {
        return new SeekBar.OnSeekBarChangeListener() {
            @Override
            public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
            }

            @Override
            public void onStartTrackingTouch(SeekBar seekBar) {
            }

            @Override
            public void onStopTrackingTouch(SeekBar seekBar) {
                viewModel.seekTo(seekBar.getProgress());
            }
        };
    }

    private String getStringSongDuration(long songDuration) {
        int hours = (int) TimeUnit.MILLISECONDS.toHours(songDuration);
        int minutes = (int) (TimeUnit.MILLISECONDS.toMinutes(songDuration)
                - TimeUnit.HOURS.toMinutes(hours));
        int seconds = (int) (TimeUnit.MILLISECONDS.toSeconds(songDuration)
                - TimeUnit.MINUTES.toSeconds(minutes)
                - TimeUnit.HOURS.toSeconds(hours));

        return convertTimeFormat(hours, minutes, seconds);
    }

    private String convertTimeFormat(int hours, int minutes, int seconds) {
        String oldFormat = "m:s";
        String newFormat = "mm:ss";
        String stringSongDuration = MessageFormat.format("{0}:{1}", minutes, seconds);

        if (hours > 0) {
            oldFormat = "h:m:s";
            newFormat = "hh:mm:ss";
            stringSongDuration = MessageFormat.format("{0}:{1}:{2}", hours, minutes, seconds);
        }

        try {
            stringSongDuration = new SimpleDateFormat(newFormat, Locale.getDefault())
                    .format(new SimpleDateFormat(oldFormat, Locale.getDefault()).parse(stringSongDuration));
        } catch (ParseException e) {
            e.printStackTrace();
        }

        return stringSongDuration;
    }

    private SharedPreferences getPreference() {
        return requireActivity().getSharedPreferences(getString(R.string.authorization_info), Context.MODE_PRIVATE);
    }

    private ViewModelProvider.Factory getFactory() {
        String userAgent = getPreference().getString(getResources().getString(R.string.user_agent), "");
        String token = getPreference().getString(getResources().getString(R.string.token_key),"");
        //Добавть проерку userAgent ибо малоли что может произойти перебздеть стоит.
        return Injection.INSTANCE.provideViewModelFactory(userAgent,token,requireContext());
    }
}

File: app/src/main/java/com/lis/player_java/ui/PlayerFragment.kt
package com.lis.player_java.ui

import android.content.Context
import android.content.SharedPreferences
import android.icu.text.SimpleDateFormat
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.SeekBar
import android.widget.SeekBar.OnSeekBarChangeListener
import android.widget.Toast
import androidx.fragment.app.Fragment
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.RecyclerView
import com.google.android.exoplayer2.Player
import com.lis.player_java.R
import com.lis.player_java.data.room.MusicDB
import com.lis.player_java.databinding.FragmentPlayerBinding
import com.lis.player_java.di.Injection
import com.lis.player_java.tool.ImageFun
import com.lis.player_java.viewModel.PlaybackViewModel
import java.text.MessageFormat
import java.text.ParseException
import java.util.*
import java.util.concurrent.TimeUnit

private var Instant: PlayerFragment? = null

fun PlayerFragment.create(): PlayerFragment {
    if (Instant == null) {
        Instant = this
    }
    return Instant!!
}

class PlayerFragment : Fragment() {
    private lateinit var binding: FragmentPlayerBinding
    private lateinit var viewModel: PlaybackViewModel

    val musicId: MutableLiveData<Long> = MutableLiveData()
    val recyclerView: MutableLiveData<RecyclerView> = MutableLiveData()

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        binding = FragmentPlayerBinding.inflate(inflater, container, false)
        viewModel = ViewModelProvider(
            this, factory
        )[PlaybackViewModel::class.java]
        binding.bindElement()
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

    }

    fun changeMusic() {
        musicId.observe(viewLifecycleOwner) { id ->
            viewModel.currentSong.value = id
        }
        recyclerView.observe(viewLifecycleOwner) {
            binding.musicList.adapter = it.adapter
        }
    }

    private fun FragmentPlayerBinding.bindElement() {
        viewModel.duration.observe(viewLifecycleOwner) { songDuration: Long ->
            Toast.makeText(requireContext(), "" + songDuration, Toast.LENGTH_SHORT).show()
            songProgress.max = songDuration.toInt()
            this.songDuration.text = getStringSongDuration(songDuration)
        }
        val imageFun = ImageFun()
        viewModel.musicInfo.observe(viewLifecycleOwner) { musicInfo: MusicDB? ->
            if (musicInfo != null) {
                val image = musicInfo.photo1200
                if (image.isNotEmpty()) {
                    imageFun.setImage(songImage, image)
                    imageFun.setImageToBackground(backgroundImage, image)
                } else {
                    imageFun.setImage(songImage, R.drawable.song_image)
                    imageFun.setImageToBackground(backgroundImage, R.drawable.song_image)
                }
                songName.text = musicInfo.title
                songName.isSelected = true
                songAuthor.text = musicInfo.artist
            }
        } //заполнение сведений о песне
        viewModel.position.observe(viewLifecycleOwner) { position: Long ->
            songPosition.text = getStringSongDuration(position)
            songProgress.progress = position.toInt()
        } //установка текущей позиции
        viewModel.downloadPosition.observe(viewLifecycleOwner) { downloadPosition: Long ->
            songProgress.secondaryProgress = downloadPosition.toInt()
        }
        viewModel.isPlaying.observe(viewLifecycleOwner) { isPlaying: Boolean ->
            imageFun.setImage(
                buttonPlayPause,
                if (isPlaying) R.drawable.pause else R.drawable.play
            )
        } //установка иконки play\pause кнопки

        viewModel.repeatMode.observe(viewLifecycleOwner) { repeatMode: Int ->
            when (repeatMode) {
                Player.REPEAT_MODE_OFF -> imageFun.setImage(
                    buttonLoop, R.drawable.repeate_mode_one
                )
                Player.REPEAT_MODE_ONE -> imageFun.setImage(
                    buttonLoop,
                    R.drawable.repeate_mode_all
                )
                Player.REPEAT_MODE_ALL -> imageFun.setImage(
                    buttonLoop,
                    R.drawable.repeat_mode_off
                )
            }
        } // установка иконки зацикливания
        viewModel.count.observe(viewLifecycleOwner) { count: Int ->
            Toast.makeText(
                requireContext(),
                "" + count,
                Toast.LENGTH_SHORT
            ).show()
            //TODO("delete this")
        }
        songProgress.setOnSeekBarChangeListener(seekBarSelectProgressListener())
        buttonPlayPause.setOnClickListener { startClickListener() }
        buttonNext.setOnClickListener { nextClickListener() }
        buttonPrevious.setOnClickListener { prevClickListener() }
        buttonLoop.setOnClickListener { loopClickListener() }
        buttonPlaylist.setOnClickListener { openPlaylist(it) }
    }

    private fun openPlaylist(view: View) {
        if (view.visibility == View.VISIBLE) {
            view.visibility = View.GONE
        } else if (view.visibility == View.GONE) {
            view.visibility = View.VISIBLE
        }

    }

    private var repeatMode = Player.REPEAT_MODE_OFF
    private fun loopClickListener() {
        when (repeatMode) {
            Player.REPEAT_MODE_OFF -> repeatMode = Player.REPEAT_MODE_ONE
            Player.REPEAT_MODE_ONE -> repeatMode = Player.REPEAT_MODE_ALL
            Player.REPEAT_MODE_ALL -> repeatMode = Player.REPEAT_MODE_OFF
        }
        viewModel.setRepeatMode(repeatMode)
    }

    private fun startClickListener() {
        var isPlaying = java.lang.Boolean.TRUE == viewModel.isPlaying.value
        if (isPlaying) {
            viewModel.pause()
        } else {
            viewModel.play()
        }
        isPlaying = java.lang.Boolean.TRUE == viewModel.isPlaying.value
    }

    private fun nextClickListener() {
        viewModel.nextSong()
    }

    private fun prevClickListener() {
        viewModel.prevSong()
    }

    private fun seekBarSelectProgressListener(): OnSeekBarChangeListener {
        return object : OnSeekBarChangeListener {
            override fun onProgressChanged(seekBar: SeekBar, progress: Int, fromUser: Boolean) {}
            override fun onStartTrackingTouch(seekBar: SeekBar) {}
            override fun onStopTrackingTouch(seekBar: SeekBar) {
                viewModel.seekTo(seekBar.progress.toLong())
            }
        }
    }

    private fun getStringSongDuration(songDuration: Long): String {
        val hours = TimeUnit.MILLISECONDS.toHours(songDuration).toInt()
        val minutes = (TimeUnit.MILLISECONDS.toMinutes(songDuration)
                - TimeUnit.HOURS.toMinutes(hours.toLong())).toInt()
        val seconds = (TimeUnit.MILLISECONDS.toSeconds(songDuration)
                - TimeUnit.MINUTES.toSeconds(minutes.toLong())
                - TimeUnit.HOURS.toSeconds(hours.toLong())).toInt()
        return convertTimeFormat(hours, minutes, seconds)
    }

    private fun convertTimeFormat(hours: Int, minutes: Int, seconds: Int): String {
        var oldFormat = "m:s"
        var newFormat = "mm:ss"
        var stringSongDuration = MessageFormat.format("{0}:{1}", minutes, seconds)
        if (hours > 0) {
            oldFormat = "h:m:s"
            newFormat = "hh:mm:ss"
            stringSongDuration = MessageFormat.format("{0}:{1}:{2}", hours, minutes, seconds)
        }
        try {
            stringSongDuration = SimpleDateFormat(newFormat, Locale.getDefault())
                .format(SimpleDateFormat(oldFormat, Locale.getDefault()).parse(stringSongDuration))
        } catch (e: ParseException) {
            e.printStackTrace()
        }
        return stringSongDuration
    }

    private val preference: SharedPreferences
        get() = requireActivity().getSharedPreferences(
            getString(R.string.authorization_info),
            Context.MODE_PRIVATE
        )

    //Добавть проерку userAgent ибо малоли что может произойти перебздеть стоит.
    private val factory: ViewModelProvider.Factory
        get() {
            val userAgent = preference.getString(resources.getString(R.string.user_agent), "")
            val token = preference.getString(resources.getString(R.string.token_key), "")
            //Добавть проерку userAgent ибо малоли что может произойти перебздеть стоит.
            return Injection.provideViewModelFactory(userAgent!!, token!!, requireContext())
        }
}